### ISSUE TITLE
Proposal: presenting invalid options

### ISSUE THREAD

**Issue by McJones** [YarnSpinnerTool/YarnSpinner#265]
# Rationale

It's not uncommon in dialogue heavy games to show dialogue options that are locked out due to having not met certain requirements.
In Mass Effect for example you need certain levels of Paragon/Renegade to be able to convice Saren of him being indoctrinated or to stop the Illusive Man.
Or in Fallout if you've high (or low) enough levels of different skill you can choose to bypass conversation encounters.

Currently in YS while we can do conditional shortcuts these are used to prune the potential options.
It is desirable to be able to both have the option shown but also have the conditional determine its viability.

This would mean a change that in dialogue like the following:

```
TD-110: Let me see your identification.
-> Of course... um totally not General Kenobi and the son of Darth Vader.
    Luke: Wait, what?!
    TD-110: Promotion Time!
-> You don't need to see his identification. <<if $learnt_mind_trick is true>>
    TD-110: We don't need to see his identification.
```

regardless of what `$learnt_mind_trick` is you'd still be sent both, just one would be flagged as invalid.
Then both could be shown, the second one could then be greyed out or some similar effect if invalid.

# The Proposal

This proposal suggests that we modify the compiler code emission and virtual machine around options to support the concept of valid and invalid options.
In the compiler currently we emit code that runs through the conditional and sets up a jump over adding the new options should the conditional fail.

We'd have to modify this so instead of jumping over the option we instead flag it as invalid.
Potential way to do this is to modify the operands for the AddOption opCode instead of having a line and number of substitutions of the line we make it so this opCode has a line, a valid/invalid flag, and then the number of substitutions.

From the user space this would mean a change from receiving a collection of options as they currently to exist to either adding a new field to each option sent over, or send over a collection of (option, validity) tuples.

# Presentation

The default views in Unity Yarn Spinner would have to be updated to skip over any invalid options.
Would likely also need to create a view that supports the presentation of invalid options so that they can be shown how to be handled.

## Informing the user

Many games that have conditional dialogue often show the reason why you can't select that option.
I think this should be doable using the format functions and markup to allow for the presentation to show the reasoning why when the conditional is false, and otherwise just appear as a regular option.

# Disadvantages

This requires modifying the opCodes, compiler, virtual machine, and dialogue views.
It touches literally EVERY part of the Yarn Spinner stack which is a risky thing to do.
As described here it also requires every person modify their dialogue even if they aren't making use of it.

# Alternatives

Could do nothing, while this does mean the feature is unsupported there have only been a few requests for this so it isn't a hugely requested feature.

Hashtags could be used here instead, could define a hashtag for validity (eg `#valid`) and then the variable to check against (eg `#valid $learnt_mind_trick`).
This means it can be instead handled all in the users space instead of then using the conditional option you just use normal options and use the hashtag to work out if it is valid or not.

**Comment by desplesda** [comment]
This is something multiple users have expressed an interest in.

I'd love to see a proposed syntax for this!

**Comment by desplesda** [comment]
Oh, I just re-read your proposal! You're suggesting no change to the syntax, but a change to the semantics of what a failing condition at the end of an option means. Do I have that right?

**Comment by McJones** [comment]
Yeah, no syntax change necessary as it reuses the already existing shortcuts with conditionals, just changes literally every other part of Yarn Spinner...

**Comment by desplesda** [comment]
Right!

In cases where a game might need to differentiate between three states (valid option, invalid option, hidden option), we could encourage them to add a hashtag on options that should be hidden if their condition fails, and detect the presence of that hashtag in their dialogue views.

**Comment by desplesda** [comment]
Actually - what we could do in an upgrader is add such a hashtag to lines, so that existing behaviour is preserved.

The following 1.0 code:

```
-> You don't need to see his identification. <<if $learnt_mind_trick is true>>
```

Could become:

```
-> You don't need to see his identification. <<if $learnt_mind_trick is true>> #hide_if_false
```

**Comment by unknowndevice** [comment]
I think this is worth doing. It's a common thing people ask for.

**Comment by desplesda** [comment]
Also, I'd like to recognise @McJones' correct identification of the stormtrooper who Obi-Wan mind-tricks in Star Wars.

**Comment by McJones** [comment]
> Also, I'd like to recognise @McJones' correct identification of the stormtrooper who Obi-Wan mind-tricks in Star Wars.

Thank you, it's the little details like those that, in my mind at least, make the difference between a good GitHub issue and an ok one.

**Comment by WeslomPo** [comment]
In my language I added symbol `!` - that represent `loud`, so if you write condition like 
`-> text | nodeId ~!if (condition)` - this option will be marked as loud - and result of this condition will be recorded to field of option (so, silent will be like `-> text | nodeId ~if (condition)`).
Also that you can mark as loud any expression - like `"~{1 + 1}" and "~!{2 + 2}"` - will write `"" and "4"`

**Comment by sh1mmer** [comment]
> Actually - what we could do in an upgrader is add such a hashtag to lines, so that existing behaviour is preserved.

I was also looking for a way to do shown but disabled dialog options. I like @desplesda's suggestion because I think you could use it for various presentation treatments not just disabled options. For example, the UI layer could use it do bold or italics treatments.

I was playing around with adding markup to the beginning of dialog options I wanted similar to @WeslomPo but supporting tags seems like it would be easier.

Maybe something like `#bold:true` or `#bold:{$foo}` would work well where tags are just K/V pairs that the UI layer can get as an array/hash and then decide what to do with.

**Comment by desplesda** [comment]
This is now implemented as of d482dd18715d7aa4e65a840161a3ca2ed36dbef5. I've also updated YarnSpinner-Unity to allow it to make use of the 'is option available' flag that's delivered to the game.

Thanks to @McJones for proposing this!
